Add code example to Potential docstrings

Dhruvanshu-Joshi opened the pull request:
<!-- !! Thank your for opening a PR !! -->

**What is this PR about?**
In this PR , I have provided a solution to close the issue [#6399](https://github.com/pymc-devs/pymc/issues/6399).
The docstring for the Potential function is rather limited. I have made efforts to include some code examples to better understand this. I also focused on the nature of the arguments that are passed to this function and have included an example for the same.


**Checklist**
+ [ ] Explain important implementation details 👆
+ [ ] Make sure that [the pre-commit linting/style checks pass](https://docs.pymc.io/en/latest/contributing/python_style.html).
+ [x] Link relevant issues (preferably in [nice commit messages](https://tbaggery.com/2008/04/19/a-note-about-git-commit-messages.html))
+ [ ] Are the changes covered by tests and docstrings?
+ [x] Fill out the short summary sections 👇

## Documentation
- Add code example and a short explanation of the code example to Potential docstrings.
- In the first example, we define a constraint such that the probability density that the model produces takes into consideration that the sum of x and y should be lesser than 1.
- In the second example, we emphasis that the var argument can be a PyTensor or Scalar variable. For the same purpose, I use a function that'll return a Scalar variable. Had we not done it , we might have encountered an error.

Armavica committed:
Move joint_logprob to test/logprob/utils.py

twiecki committed:
Allow external nuts sampler directly from `sample`

Co-authored-by: Ricardo Vieira <28983449+ricardov94@users.noreply.github.com>

Armavica committed:
Move pymc/distributions/logprob.py to pymc/logprob

* _joint_logp to pymc/logprob/joint_logprob.py
* _get_scaling, _check_no_rvs, logp to logprob/joint_logprob.py
* logcdf to logprob/abstract.py
* ignore_logprob to logprob/utils.py

michaelosthege committed:
Introduce `IBaseTrace` with minimal API for chain backends

michaelosthege committed:
Introduce `IBaseTrace` with minimal API for chain backends

michaelosthege committed:
Modify internal sampling functions from `BaseTrace` to `IBaseTrace`

michaelosthege committed:
Modify internal sampling functions from `BaseTrace` to `IBaseTrace`

michaelosthege committed:
Create traces with `init_traces` helper function

michaelosthege committed:
Create traces with `init_traces` helper function

michaelosthege committed:
Fix `test_weibull_logp` flakyness

michaelosthege committed:
Fix `test_weibull_logp` flakyness

Dhruvanshu-Joshi committed:
Docstring for Potential function updated

Dhruvanshu-Joshi committed:
Docstring for Potential function updated

larryshamalama commented on pymc/model.py:
```diff
@@ -2035,10 +2035,43 @@ def Deterministic(name, var, model=None, dims=None):
 def Potential(name, var, model=None):
     """Add an arbitrary factor potential to the model likelihood
 
+      Potential function is used to add arbitrary factors i.e. constraints that we define to the likelihood function of a model to adjust the probablity density of the model. 
```
```suggestion
      The Potential function is used to add arbitrary factors (such as constraints or other likelihood components) to adjust the probability density of the model. 
```

larryshamalama commented on pymc/model.py:
```diff
@@ -2035,10 +2035,43 @@ def Deterministic(name, var, model=None, dims=None):
+
+        with pm.Model() as model:
+          x = pm.Normal('x', mu=0, sigma=1)
+          y = pm.Normal('y',mu=x, sigma=1,observed=data)
```
```suggestion
          y = pm.Normal('y', mu=x, sigma=1, observed=data)
```

larryshamalama commented on pymc/model.py:
```diff
@@ -2035,10 +2035,43 @@ def Deterministic(name, var, model=None, dims=None):
+          constraint = x + y < 1
+          potential = pm.Potential('constraint', constraint, my_model)
+
+        In this example, we define a constraint on the sum of x and y that it can only be lesser than 1. We pass this constraint to the pm.Potential function.
```
```suggestion
        In this example, we define a constraint on the sum of x and y to be less than 1. We pass this constraint to the ``pm.Potential`` function.
```

larryshamalama commented on pymc/model.py:
```diff
@@ -2035,10 +2035,43 @@ def Deterministic(name, var, model=None, dims=None):
+          potential = pm.Potential('constraint', constraint, my_model)
+
+        In this example, we define a constraint on the sum of x and y that it can only be lesser than 1. We pass this constraint to the pm.Potential function.
+        Now the pm.Potential function takes in a expression, a scalar value and the model name as argument.
```
```suggestion
```

I would remove this line because the `Parameters` section below provides this information.

ricardoV94 replied:
I agree with @larryshamalama. Also note that it doesn't have to return a scalar expression at all. Whatever is returned will be summed into a scalar automatically, so a user can return an arbitrary shaped expression

larryshamalama commented on pymc/model.py:
```diff
@@ -2035,10 +2035,43 @@ def Deterministic(name, var, model=None, dims=None):
+
+        In this example, we define a constraint on the sum of x and y that it can only be lesser than 1. We pass this constraint to the pm.Potential function.
+        Now the pm.Potential function takes in a expression, a scalar value and the model name as argument.
+        Now, the probablity density that the model produces takes into consideration the constraint that the sum of x and y should be lesser than 1.
```
```suggestion
        Now, the probability density that the model produces takes into consideration the constraint that the sum of x and y should be less than 1.
```

larryshamalama commented on pymc/model.py:
```diff
@@ -2035,10 +2035,43 @@ def Deterministic(name, var, model=None, dims=None):
+          x = pm.Normal('x', mu=0, sigma=1)
+          y = pm.Normal('y',mu=x, sigma=1,observed=data)
+          constraint = x>= 0
+          potential = pm.Potential('x_constraint',-tt.log(tt.switch(constraint, 1, 1e-30)))
```
```suggestion
          potential = pm.Potential('x_constraint', -pt.log(pt.switch(constraint, 1, 1e-30)))
```

We recently switched our backend to [PyTensor](https://github.com/pymc-devs/pytensor)!

Dhruvanshu-Joshi replied:
If we are avoiding to use too many new concepts at once, can we use pm.math for backend computation here instead of explicitly mentioning Theano tensor or PyTensor?
I am quite new to PyTensor and am facing errors when I try to run the suggestion made above.

larryshamalama replied:
Yes, that's a great idea. @ricardoV94 mentioned it below too

larryshamalama commented on pymc/model.py:
```diff
@@ -2035,10 +2035,43 @@ def Deterministic(name, var, model=None, dims=None):
     name: str
-    var: PyTensor variables
+    var: PyTensor or Scalar variables with the same shape as input to the likelihood
+    model: The model object to which the potential function is added. If none is provided, current model is used
```
```suggestion
    model: The model object to which the potential function is added. If ``None`` is provided, the current model is used.
```

larryshamalama commented on pymc/model.py:
```diff
@@ -2035,10 +2035,43 @@ def Deterministic(name, var, model=None, dims=None):
+          constraint = x>= 0
+          potential = pm.Potential('x_constraint',-tt.log(tt.switch(constraint, 1, 1e-30)))
+
+        In this example, we define a constraint on x that it can be either greater than or equal to 0 and pass this constraint to the pm.Potential function.
```
```suggestion
        In this example, we define a constraint on x to be greater or equal to 0 via the ``pm.Potential`` function.
```

larryshamalama commented on pymc/model.py:
```diff
@@ -2035,10 +2035,43 @@ def Deterministic(name, var, model=None, dims=None):
+          potential = pm.Potential('x_constraint',-tt.log(tt.switch(constraint, 1, 1e-30)))
+
+        In this example, we define a constraint on x that it can be either greater than or equal to 0 and pass this constraint to the pm.Potential function.
+        Now the pm.Potential function takes in a string expression and a scalar value as argument. Hence we pass '-tt.log(tt.switch(constraint, 1, 1e-30))' as second argument which will return a scalar value depending on the constraint and will add to the likelihood function of the model.
```
```suggestion
        We pass ``-pt.log(pt.switch(constraint, 1, 1e-30))`` as second argument which will return a scalar value depending on if the constraint is met or not and which will added to the likelihood of the model.
```

larryshamalama commented on pymc/model.py:
```diff
@@ -2035,10 +2035,43 @@ def Deterministic(name, var, model=None, dims=None):
+
+        In this example, we define a constraint on x that it can be either greater than or equal to 0 and pass this constraint to the pm.Potential function.
+        Now the pm.Potential function takes in a string expression and a scalar value as argument. Hence we pass '-tt.log(tt.switch(constraint, 1, 1e-30))' as second argument which will return a scalar value depending on the constraint and will add to the likelihood function of the model.
+        This is done because PyMC3 uses an energy-based model, which means that the probability of a variable configuration is proportional to the negative of the energy of that configuration. So, the negative logarithm of the switch function ensures that the energy is high (i.e. close to negative infinity) when the constraint is not met, and low (i.e. close to zero) when the constraint is met.
```
```suggestion
The negative logarithm of the switch function ensures that this component of the log-likelihood is high (i.e. close to 30) when the constraint is not met, and low (i.e. close to zero) when the constraint is met.
```

I'm confused with this sentence. What do you mean by "probability of a variable configuration"?

Lastly, personally, I would avoid using the energy concept since it's similar to the likelihood and focus on the log-likelihood since that's what PyMC works with. In the sense that I would avoid introducing too many new concepts as once, although the energy-based analogy is correct (to my understanding and despite my little knowledge of physics 😅). 

However, with that, the sentence would revert the direction of the logic if I'm not mistaken...

Dhruvanshu-Joshi replied:
Answering the first question,  'probability of variable configuration' refers to the likelihood function. By this I simply  mean to refer the likelihood of a particular set of values for the variables that the model equation uses.

For the second part, no it does not revert the direction of logic... That's bad framing on my part😅

What I meant from "close to  negative infinity" was that the log function/ log-likelihood will be close to negative infinity. The mathematical negative will thus be a large positive value. Energy is the negative logarithmic function that we have defined and thus it will be high when the constraint is not met. 

Can we re-frame this as:
The reciprocal exponential function of the negative logarithm is proportional to the likelihood function. When the constraint is met, the switch function returns 1 making the negative logarithm 0 and hence likelihood function has no penalizing effect. When the constraint is not met the switch function returns 1e-30 and the negative logarithm takes a high value(i.e. 30), reciprocal exponential function of which makes the likelihood function very small and successfully penalizes it.

larryshamalama replied:
Sorry, chipping after much time. I'm not sure what you mean by "reciprocal exponential function of the negative logarithm"... 

However, it looks like you have it figured out in some threads below! Perhaps you can focus on those suggestions rather than on my comments

larryshamalama reviewed (changes requested):
Hi @Dhruvanshu-Joshi, thanks for creating this pull request! Also, thanks for your patience, it seems that our devs have been busy recently... (myself included)

Overall, it looks good! I gave some suggestions to tweak some sentences here and there. Let me know if you would like any assistance with this PR.

Edit: It seems that the pre-commit has not been run successfully; let us know if you would like any pointers regarding that.

ricardoV94 commented on pymc/model.py:
```diff
@@ -2035,10 +2035,43 @@ def Deterministic(name, var, model=None, dims=None):
+          y = pm.Normal('y', mu=x, tau=1)
+
+          constraint = x + y < 1
+          potential = pm.Potential('constraint', constraint, my_model)
```
This example does not make sense like this. x + y < 1 is gonna return a variable that evaluates to 0 or 1, which will be added to the model logp. It would make sense to add the `log(constraint)` instead. 

However, I think it's a bit more readable for users to see a constraint with a switch as that's usually how it crops up in examples. Something like `pm.Potential("constraint", pt.log(pt.switch(constraint, 1, 0)))`

But this is very similar to the example below, so just show one of them

Dhruvanshu-Joshi replied:
I understand how using the logarithm function makes more sense and also notice the repetitiveness in both the examples.
I will add a commit as soon as it is clear if we are moving forward with Pytensor or using pm.math.

ricardoV94 replied:
`pm.math` is fine

ricardoV94 commented on pymc/model.py:
```diff
@@ -2035,10 +2035,43 @@ def Deterministic(name, var, model=None, dims=None):
+        Now the pm.Potential function takes in a expression, a scalar value and the model name as argument.
+        Now, the probablity density that the model produces takes into consideration the constraint that the sum of x and y should be lesser than 1.
+
+        The Potential function here acts as a gatekeeper allowing only samples that satisfy the constraint to contribute to the likelihood. This ensures that the sample generated by the sampler are consistent with the constraint.
```
```suggestion
        The Potential function here acts as a gatekeeper allowing only samples that satisfy the constraint to contribute to the likelihood. This ensures that the samples generated by the sampler are consistent with the constraint.
```

ricardoV94 commented on pymc/model.py:
```diff
@@ -2035,10 +2035,43 @@ def Deterministic(name, var, model=None, dims=None):
     ----------
     name: str
-    var: PyTensor variables
+    var: PyTensor or Scalar variables with the same shape as input to the likelihood
```
```suggestion
    var: PyTensor expression to be added to the model joint logp 
```

ricardoV94 commented on pymc/model.py:
```diff
@@ -2035,10 +2035,43 @@ def Deterministic(name, var, model=None, dims=None):
+          y = pm.Normal('y', mu=x, tau=1)
+
+          constraint = x + y < 1
+          potential = pm.Potential('constraint', constraint, my_model)
```
I wouldn't show the `model` argument being used. Most times it's not needed as users add a Potential inside a model context.

ricardoV94 reviewed (changes requested):
Thanks for opening up the PR. This looks nice already. I have three notes:

## First

I think the two examples should be merged into a single one since they both show an example of using Potential to impose a constraint. The first example was incorrect though, and the second one looks better if you make it strict `pt.log(pt.switch(cond, 1, 0))` instead of the arbitrary large or small values.

I would add a second example showing how Potential can be used to obtain an arbitrary prior and not just a binary constraint. Maybe, something like this which corresponds to a power law prior for `max_items` 

```python
with pm.Model():
  # p(max_items) = 1 / max_items
  max_items = pm.Uniform("max_items", lower=1, upper=100)
  pm.Potential('power_prior', pm.math.log(1 / max_items))

  n_items = pm.Uniform('n_items', lower=1, upper=max_items, observed=60)
```

## Second
I would use `pm.math.foo` instead of `pt.foo`.

## Third
We should add a warning box saying that Potentials are only added to the model logp, and as such can only influence logp based sampling, like the one used by `pm.sample`.

They do not affect forward sampling, which is used by `sample_prior_predictive` and `sample_posterior_predictive`. A warning is always raised to alert user of this.

ricardoV94 committed:
Remove experimental module warning from jax.py

ricardoV94 committed:
Raise error when trying to jaxify graphs with RNGs

Dhruvanshu-Joshi committed:
Updated model.py with all the suggestions

Dhruvanshu-Joshi committed:
Updated model.py with all the suggestions

alekracicot committed:
Add nu parametrization to beta distribution (#6344)

* Add (mu, nu) parametrization

* add test beta logp

* Add relationship between alpha, beta and nu

* change test structure

* remove white space accidently added in docstring

* remove sample size from docstring

---------

Co-authored-by: Michael Osthege <michael.osthege@outlook.com>

alekracicot committed:
Add nu parametrization to beta distribution (#6344)

* Add (mu, nu) parametrization

* add test beta logp

* Add relationship between alpha, beta and nu

* change test structure

* remove white space accidently added in docstring

* remove sample size from docstring

---------

Co-authored-by: Michael Osthege <michael.osthege@outlook.com>

michaelosthege committed:
Allow for `StatsBijection.rmap` of incomplete stat dicts

OriolAbril reviewed (changes requested):
Currently on my mobile and can't properly review but want to leave a reminder for myself to review when I gat the chance. It should follow numpydoc: https://numpydoc.readthedocs.io/en/latest/format.html. Examples in their own section below parameters and returns. Warnings also as their own section, the spaces both before and after the parameter name and the parameter info... https://pymc-data-umbrella.xyz/en/latest/sprint/tutorials/docstring_tutorial.html#edit-the-docstring might also be helpful

ricardoV94 commented on pymc/model.py:
```diff
@@ -2035,10 +2035,49 @@ def Deterministic(name, var, model=None, dims=None):
+          constraint = x>= 0
+          potential = pm.Potential('x_constraint', -pm.math.log(pm.math.switch(constraint, 1, 1e-30))
+
+        In this example, we define a constraint on x to be greater or equal to 0 via the ``pm.Potential`` function.
```
I would put the explanations above the respective examples

ricardoV94 commented on pymc/model.py:
```diff
@@ -2035,10 +2035,49 @@ def Deterministic(name, var, model=None, dims=None):
+          x = pm.Normal('x', mu=0, sigma=1)
+          y = pm.Normal('y', mu=x, sigma=1, observed=data)
+          constraint = x>= 0
+          potential = pm.Potential('x_constraint', -pm.math.log(pm.math.switch(constraint, 1, 1e-30))
```
I think the absolute constraint with (1 vs 0) may be more educative than (1 vs 1e-30). I would perhaps show that and mention an alternative something like (1 vs 0.5) would provide a soft constraint instead of a hard one.

I find the 1e-30 odd because it behaves like 0 in practice. The sampler will never accept a negative value for 'x', or perhaps once in a lifetime, which is an odd modelling choice.

We should change the mu of x to be 0.1 perhaps, so that the sampler doesn't start at an invalid point.

Dhruvanshu-Joshi replied:
I understand that the absolute constraint with (1 vs 0) might be more educative than (1 vs 1e-30), but the problem with it is that it produces error when trying to sample the model because the logarithm of 0 is not defined. If having a working model that can be sampled without encountering errors along with explaining the use of Potential function is not what is required with this code block, a constraint of 1 vs 0 absolutely makes more sense. So to reconfirm, should I change the constraint?

Soft constraints also are very helpful. Should I make a  separate code-block for it or a commented line in the original code-block with an explanation?

ricardoV94 replied:
Logarithm of zero is well defined (at least in PyMC land) and leads to -inf logp which is exactly what you want to represent an impossible event in log scale. We use -inf explicitly in most of the logp definitions 9f distributions with constrained support.

About the soft constrained, whichever way you think is better. I have no preference.

Dhruvanshu-Joshi replied:
When I try to sample using potential = pm.Potential('x_constraint',-pm.math.log(pm.math.switch(constraint, 1, 0.0))) I face this error:

AttributeError                            Traceback (most recent call last)
[<ipython-input-31-e645e4d4afc7>](https://localhost:8080/#) in <module>
     11   potential = pm.Potential('x_constraint',-pm.math.log(pm.math.switch(constraint, 1, 0)))
     12 
---> 13   trace = pm.sample(1000,tune=1000)

12 frames
[/usr/local/lib/python3.8/dist-packages/aesara/tensor/elemwise.py](https://localhost:8080/#) in transform(r)
    600         def transform(r):
    601             # From a graph of ScalarOps, make a graph of Broadcast ops.
--> 602             if isinstance(r.type, (NullType, DisconnectedType)):
    603                 return r
    604             if r in scalar_inputs:

AttributeError: 'float' object has no attribute 'type'

On defining the Potential function as potential = pm.Potential('x_constraint',-pm.math.log(pm.math.switch(constraint, 1.0, 0.0))) I get this:

Interrupted
---------------------------------------------------------------------------
AttributeError                            Traceback (most recent call last)
[<ipython-input-34-ed86f6db8804>](https://localhost:8080/#) in <module>
     11   potential = pm.Potential('x_constraint',-pm.math.log(pm.math.switch(constraint, 1.0, 0.0)))
     12 
---> 13   trace = pm.sample(1000,tune=1000)

8 frames
[/usr/local/lib/python3.8/dist-packages/pymc/step_methods/hmc/base_hmc.py](https://localhost:8080/#) in astep(self, q0)
    166             message_energy = (
    167                 "Bad initial energy, check any log probabilities that "
--> 168                 "are inf or -inf, nan or very small:\n{}".format(error_logp.to_string())
    169             )
    170             warning = SamplerWarning(

AttributeError: 'numpy.ndarray' object has no attribute 'to_string'

ricardoV94 replied:
For it to have a valid initial energy with the hard constraint you have to set the prior of x to not be centered at zero like `x = pm.Normal("x", mu=0.1, sigma=1)` or set an initial value like `x = pm.Normal("x", mu=0, sigma=1, initval=0.1)`

Dhruvanshu-Joshi replied:
The negative sign was the problem. I think this is because if we take the negative sign, the Potential function returns in inf instead of -inf. From my understanding, the potential term should be somewhere close to 0 when the constraint is satisfied and very large when the constraint is not satisfied. What I thought was that both inf and -inf can work as this large value. But when Potential value becomes inf, for some reason I face the error "'numpy.ndarray' object has no attribute 'to_string'". So, is it wrong if the Potential value becomes inf?

ricardoV94 replied:
The potential value is just added to the model logp, so it should be -inf (or very negative) when a constraint is violated, so that those draws are rejected. 0 won't have any effect and positive values will make the proposals more likely to be accepted

Dhruvanshu-Joshi replied:
Alright👍 
I shall include this in the docstring too.
Thanks for the help!

ricardoV94 reviewed (commented):
Do you think abbreviate a bit the explanations? I don't know if it's just me but they feel a tad too long.

michaelosthege committed:
More robust tune stat fetching and type hints

The mcmc module relies on the `"tune"` stat to figure out the number of
tune/draw iterations post sampling.
These changes remove reliance on any weird squeeze-cat magic.

Co-authored-by: Virgile Andreani <armavica@ulminfo.fr>

Commit:
⬆️ UPGRADE: Autoupdate pre-commit config

ricardoV94 commented on pymc/model.py:
```diff
@@ -2035,10 +2035,49 @@ def Deterministic(name, var, model=None, dims=None):
+          x = pm.Normal('x', mu=0, sigma=1)
+          y = pm.Normal('y', mu=x, sigma=1, observed=data)
+          constraint = x>= 0
+          potential = pm.Potential('x_constraint', -pm.math.log(pm.math.switch(constraint, 1, 1e-30))
```
I don't think it should be negated

```suggestion
          potential = pm.Potential('x_constraint', pm.math.log(pm.math.switch(constraint, 1, 1e-30))
```

ricardoV94 commented on pymc/model.py:
```diff
@@ -2035,10 +2035,49 @@ def Deterministic(name, var, model=None, dims=None):
+        with pm.Model():
+          # p(max_items) = 1 / max_items
+          max_items = pm.Uniform("max_items", lower=1, upper=100)
+          pm.Potential('power_prior', -pm.math.log(max_items))
```
They are equivalent, but maybe this one is more readable for users

```suggestion
          pm.Potential('power_prior', pm.math.log(1/max_items))
```

Dhruvanshu-Joshi committed:
Additional codeblocks added and theory reduced

Dhruvanshu-Joshi committed:
Additional codeblocks added and theory reduced

Dhruvanshu-Joshi committed:
Merge Conflicts solved

Armavica committed:
Remove test redundant with test_measurable_join_*

Armavica committed:
Remove test redundant with test_mixture.py

aleicazatti committed:
Small fix in HyperGeometric docstring

michaelosthege committed:
Add `StatsBijection.n_samplers` property

michaelosthege committed:
Introduce `stats_dtypes_shape` attribute

Closes #6503

michaelosthege committed:
Specify `stats_dtypes_shapes` for all samplers

bsenst committed:
Update GLM_linear.ipynb (#6519)

* update link to john kruschkes book
* google plus has been stopped in 2018 - remove dead link

dependabot[bot] committed:
Bump docker/setup-buildx-action from 2.4.0 to 2.4.1 (#6520)

Bumps [docker/setup-buildx-action](https://github.com/docker/setup-buildx-action) from 2.4.0 to 2.4.1.
- [Release notes](https://github.com/docker/setup-buildx-action/releases)
- [Commits](https://github.com/docker/setup-buildx-action/compare/v2.4.0...v2.4.1)

---
updated-dependencies:
- dependency-name: docker/setup-buildx-action
  dependency-type: direct:production
  update-type: version-update:semver-patch
...

Signed-off-by: dependabot[bot] <support@github.com>
Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

ricardoV94 committed:
Fix flaky `test_truncation_specialized_op`

Using empty for the observed would sometimes lead PyMC into triggering automatic imputation, whenever the trash values in empty were nan.

dependabot[bot] committed:
Bump docker/build-push-action from 3 to 4 (#6505)

Bumps [docker/build-push-action](https://github.com/docker/build-push-action) from 3 to 4.
- [Release notes](https://github.com/docker/build-push-action/releases)
- [Commits](https://github.com/docker/build-push-action/compare/v3...3b5e8027fcad23fda98b2e3ac259d8d67585f671)

---
updated-dependencies:
- dependency-name: docker/build-push-action
  dependency-type: direct:production
  update-type: version-update:semver-major
...

Signed-off-by: dependabot[bot] <support@github.com>
Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

ricardoV94 committed:
Make all but `draws` keyword-only arguments in `sample`.

Reorder arguments more logically [citation needed]

Dhruvanshu-Joshi committed:
Adding function for mutable coords (#6515)

* adding function for mutable coords

* Added coords_mutable

* add coords_mutable

* Make `coords_mutable` an explicit kwarg

* Remove reference to undefined variable

* Adding an if block

---------

Co-authored-by: Michael Osthege <michael.osthege@outlook.com>

Commit:
⬆️ UPGRADE: Autoupdate pre-commit config

michaelosthege committed:
Build docs in simplified environment (#6537)

* Build docs in simplified environment

* Fix docs build

* Remove private API section from docs

soma2000-lang committed:
BetaBinomial variance is not correctly described (#6516)

Armavica committed:
Move tests outside of code directory

OriolAbril commented on pymc/model.py:
```diff
@@ -2035,10 +2035,74 @@ def Deterministic(name, var, model=None, dims=None):
+        Potential functions only influence logp based sampling, like the one used by ``pm.sample``. Potentials, modify the log-probability of the model by adding a contribution to the logp which is used by sampling algorithms which rely on the information about the observed data to generate posterior samples.
+        Potentials are not applicable in the context of forward sampling because they don't affect the prior distribution itself, only the computation of the logp.
+        Forward sampling algorithms generate sample points from the prior distribution of the model, without taking into account the likelihood function. In other words, it does not use the information about the observed data.
+        Hence, Potentials do not affect forward sampling, which is used by ``sample_prior_predictive`` and ``sample_posterior_predictive``. A warning saying "The effect of Potentials on other parameters is ignored during prior predictive sampling" is always raised to alert user of this.
```
This should use a header following https://numpydoc.readthedocs.io/en/latest/format.html#warnings

OriolAbril commented on pymc/model.py:
```diff
@@ -2035,10 +2035,74 @@ def Deterministic(name, var, model=None, dims=None):
+          y = pm.Normal('y', mu=x, sigma=1, observed=data)
+          soft_sum_constraint = pm.Potential("soft_sum_constraint", -((x + y)**2))
+
+      The potential value is incorporated into the model log-probability, so it should be -inf (or very negative) when a constraint is violated, so that those draws are rejected. 0 won't have any effect and positive values will make the proposals more likely to be accepted
```
All this should go at the bottom of the docstring and have a section header too. Ref: https://numpydoc.readthedocs.io/en/latest/format.html#examples

OriolAbril commented on pymc/model.py:
```diff
@@ -2035,10 +2035,74 @@ def Deterministic(name, var, model=None, dims=None):
     name: str
-    var: PyTensor variables
+    var: PyTensor expression to be added to the model joint logp
+    model: The model object to which the potential function is added. If ``None`` is provided, the current model is used.
```
```suggestion
    var : PyTensor expression
        Expression to be added to the model joint logp
    model : Model, optional
        The model object to which the potential function is added. If ``None`` is provided, the current model is used.
```

a space is needed both before and after the colon (existing docstrings are generally not a good example to follow), and the first line should contain only `name : type` the description comes in the 2nd+ line and indented.

michaelraczycki committed:
increase zerosum readability issue #6459 (#6522)

* added n_zerosum_axes and added backwards compatibility for previous parameter name

* fixed typo that caused the zerosum_axes param not to be saved correctly in the new param

* adapted tests to use n_zerosum_axes

---------

Co-authored-by: Michal Raczycki <michalraczycki@macbook-pro-michal.home>

michaelraczycki committed:
issue #5791, dims & cords inference from xarray (#6514)

* added dim inference from xarray, deprecation warning and unittest for the new feature

* fixed typo in warning

* fixed accidental quotation around dim

* fixed failing assertions

* found and fixed cause of the failing test

* changed the coords assertion according to suggested form

* fixing mypy type missmatch

* working on getting the test to work

* removed typecasting to string on dim_name, was causing the mypy to fail

* took care locally of mypy errors

* Typo/formatting fixes

---------

Co-authored-by: Michal Raczycki <michalraczycki@macbook-pro-michal.home>
Co-authored-by: Michael Osthege <michael.osthege@outlook.com>

Armavica committed:
Fix some code lints found with ruff (#6545)

* Make isort happy with imports order

* Avoid shadowing the logprob fn with a local name

* Deal with unused error names

* Remove unreachable return

* Remove unread variable

* Fix type checking for variational/approximations

* A few typing fixes

michaelraczycki committed:
Add check for variables in step samplers (#6524)

* added value check for step samplers

* changing error message to be more informative

* added test checking if variable not being in model.value_vars will trigger Value error

* implemented changes from PR review

* Fix rebase

---------

Co-authored-by: Michael Osthege <michael.osthege@outlook.com>

michaelosthege committed:
Refactor sample return (#6546)

* Extract return part of `pm.sample`

* Speed up `test_mcmc`

* Group tests related to `pm.sample` return parameters

* Consolidate tests of return options

Removes a regression test added in #3821 because it took 14 seconds.

Dhruvanshu-Joshi committed:
Update Potential docstring according to suggestions

Dhruvanshu-Joshi committed:
Update docstring according to suggestions

OriolAbril reviewed (approved):
Doc preview: https://pymcio--6489.org.readthedocs.build/projects/docs/en/6489/api/generated/pymc.Potential.html looks good, I haven't reviewed the code nor looked at the potential merge/rebase issues
